src/usdb_syncer/gui/mw.py

```diff
@@ -263,6 +269,77 @@ def _generate_song_pdf(self) -> None:
         if path:
             generate_song_pdf(self.table.all_local_songs(), path)
 
+    def _import_usdb_ids_from_files(self) -> None:
+        file_list = QFileDialog.getOpenFileNames(
+            self,
+            caption="Select one or more files to import USDB IDs from",
+            dir=os.getcwd(),
+            filter="JSON, USDB IDs, Weblinks (*.json *.usdb_ids *.url *.webloc *.desktop)",
+        )[0]
+        if not file_list:
+            _logger.info("no files selected to import USDB IDs from")
+            return
+        file_parsers = [UsdbIdFile.parse(path) for path in file_list]
+        has_error = False
+        for count, parser in enumerate(file_parsers):
+            if parser.error:
+                _logger.error(
+                    f"failed importing file {file_list[count]}: {str(parser.error)}"
+                )
+                has_error = True
+        # stop import if encounter errors
+        if has_error:
+            return
+        song_ids = [id for parser in file_parsers for id in parser.ids]
+        unique_song_ids = list(set(song_ids))
+        unique_song_ids.sort()
+        _logger.info(
+            f"read {len(file_list)} file(s), "
+            f"found {len(unique_song_ids)} "
+            f"USDB IDs: {', '.join(str(id) for id in unique_song_ids)}"
+        )
+        if unavailable_song_ids := [
+            song_id for song_id in unique_song_ids if not self.table.get_data(song_id)
+        ]:
+            _logger.warning(
+                f"{len(unavailable_song_ids)}/{len(unique_song_ids)} "
+                "imported USDB IDs are not available: "
+                f"{', '.join(str(song_id) for song_id in unavailable_song_ids)}"
+            )
+        if available_song_ids := [
+            song_id
+            for song_id in unique_song_ids
+            if song_id not in unavailable_song_ids
+        ]:
+            # select available songs to prepare Download
+            _logger.info(
+                f"available {len(available_song_ids)}/{len(unique_song_ids)} "
+                "imported USDB IDs are added to Batch: "
+                f"{', '.join(str(song_id) for song_id in available_song_ids)}"
+            )
+            self.table.stage_song_ids(available_song_ids)
+
+    def _export_usdb_ids_to_file(self) -> None:
+        selected_ids = self.table.batch_ids()
```

Comment:
You have to create a list here, otherwise `not selected_ids` will always be `False`, because `selected_ids` is a lazy generator which doesn't know its length. Also, you can only use the generator once. So in line 341 it will be empty, because you've consumed it in line 339.

Reply:
... ah did not know that detail.  I spotted that I convert to a list in 339 anyways, I guess thats why conversion in 323 instead is a good solution here :-)